src/solvers/smt2_incremental/smt_to_string.cpp

```diff
@@ -0,0 +1,375 @@
+  push_outputs(std::forward<outputst>(outputs)...);
+  push_output(std::forward<outputt>(output));
+}
+
+void smt_term_to_string_convertert::visit(
+  const smt_bool_literal_termt &bool_literal)
+{
+  push_output(bool_literal.value() ? "true" : "false");
+}
+
+void smt_term_to_string_convertert::visit(const smt_not_termt &not_term)
+{
+  push_outputs("(not ", not_term.operand(), ")");
+}
+
+void smt_term_to_string_convertert::visit(
+  const smt_identifier_termt &identifier_term)
+{
+  push_outputs(
+    "|", smt2_convt::convert_identifier(identifier_term.identifier()), "|");
+}
+
+void smt_term_to_string_convertert::visit(
+  const smt_bit_vector_constant_termt &bit_vector_constant)
+{
+  auto value = integer2string(bit_vector_constant.value());
+  auto bit_width = std::to_string(bit_vector_constant.get_sort().bit_width());
+  push_outputs("(_ bv", std::move(value), " ", std::move(bit_width), ")");
+}
+
+void smt_term_to_string_convertert::visit(
+  const smt_function_application_termt &function_application)
+{
+  const auto &id = function_application.function_identifier();
+  auto arguments = function_application.arguments();
+  push_outputs("(", id, std::move(arguments), ")");
+}
+
+std::ostream &
+smt_term_to_string_convertert::convert(std::ostream &os, const smt_termt &term)
+{
+  smt_term_to_string_convertert converter;
+  term.accept(converter);
+  while(!converter.output_stack.empty())
+  {
+    auto output_function = std::move(converter.output_stack.top());
+    converter.output_stack.pop();
+    output_function(os);
+  }
+  return os;
+}
+
+std::ostream &operator<<(std::ostream &os, const smt_termt &term)
+{
+  return smt_term_to_string_convertert::convert(os, term);
+}
+
+std::string smt_to_string(const smt_termt &term)
+{
+  std::stringstream ss;
+  ss << term;
+  return ss.str();
+}
+
+class smt_option_to_string_convertert
+  : public smt_option_const_downcast_visitort
+{
+protected:
+  std::ostream &os;
+
+public:
+  explicit smt_option_to_string_convertert(std::ostream &os) : os(os)
+  {
+  }
+
+  void visit(const smt_option_produce_modelst &produce_models) override
+  {
+    os << ":produce-models " << (produce_models.setting() ? "true" : "false");
+  }
+};
+
+std::ostream &operator<<(std::ostream &os, const smt_optiont &option)
+{
+  option.accept(smt_option_to_string_convertert{os});
+  return os;
+}
+
+std::string smt_to_string(const smt_optiont &option)
+{
+  std::stringstream ss;
+  ss << option;
+  return ss.str();
+}
+
+class smt_logic_to_string_convertert : public smt_logic_const_downcast_visitort
+{
+protected:
+  std::ostream &os;
+
+public:
+  explicit smt_logic_to_string_convertert(std::ostream &os) : os(os)
+  {
+  }
+
+  void visit(const smt_logic_quantifier_free_bit_vectorst &) override
+  {
+    os << "QF_BV";
+  }
+};
+
+std::ostream &operator<<(std::ostream &os, const smt_logict &logic)
+{
+  logic.accept(smt_logic_to_string_convertert{os});
+  return os;
+}
+
+std::string smt_to_string(const smt_logict &logic)
```

Comment:
Should we make the naming here (and in the class/file name) a little more explicit? e.g. `smt_to_smt2_string` ? Otherwise its not explicitly clear what sort of 'string' is being returned.

Reply:
I am fine with making that change if you think it would make things clearer. I can't think of many other kinds of string which we could be converting to, but that is probably because I've been busy writing this one.